### PR TITLE
[b/359499526] Dump raw Ranger API entries to output jsonl files.

### PR DESCRIPTION
**Pull request by aleofreddi**

### CODE REVIEW

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java [hunk 1]
```diff
@@ -199,7 +132,19 @@ private <T> T doHttpGet(String path, Map<String, String> params, Class<T> bodyCl
...
+        }
+        Object items = payload.get(payloadProperty);
+        if (!(items instanceof List)) {
+          throw new RangerException(
+              "Expected an array value for key "
+                  + payloadProperty
+                  + " in Ranger response, got "
+                  + items.getClass());
+        }
+        return ImmutableList.copyOf((List<Object>) items);
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java:147]: Should this be a `List<?>`, not a `List<Object>`, since we don't actually know what type is in it, but that type does exist, and is fixed?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java [hunk 2]
```diff
@@ -127,14 +123,14 @@ protected String toCallDescription() {
...
     DumpGroupsTask() {
       super(GroupsFormat.ZIP_ENTRY_NAME);
     }
 
     @Override
-    protected Iterator<Group> dataIterator(@Nonnull RangerClientHandle handle) {
+    protected Iterator<Object> dataIterator(@Nonnull RangerClientHandle handle) {
       return new RangerPageIterator<>(
           page -> {
             try {
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java:136]: Should the try/catch be in RangerPagerIterator.computeNext()? It would have to use its own FunctionalInterface, but that's no cost. It would make all these callers much cleaner, and reduce the number (if not the call-frequency) of 'throw new RuntimeException' calls, which is generally an antipattern.

Might be out of the scope of this PR though, as you didn't change that.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java [hunk 3]
```diff
@@ -226,22 +222,22 @@ protected String toCallDescription() {
...
+    protected abstract Iterator<Object> dataIterator(@Nonnull RangerClientHandle handle);
 
     @Override
     protected Void doRun(TaskRunContext context, @Nonnull ByteSink sink, @Nonnull Handle handle)
         throws Exception {
       RangerClientHandle rangerClientHandler = (RangerClientHandle) handle;
       LOG.info("Writing to '{}' -> '{}'", getTargetPath(), sink);
       try (Writer writer = sink.asCharSink(StandardCharsets.UTF_8).openBufferedStream()) {
-        for (Iterator<T> iterator = dataIterator(rangerClientHandler); iterator.hasNext(); ) {
+        for (Iterator<Object> iterator = dataIterator(rangerClientHandler); iterator.hasNext(); ) {
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java:240]: I looked briefly at this to see whether it _had_ to use Iterator, and technically, we could replace this with another type/class/pattern, even a RangerPagedCollection.forEach() pattern which would not require the exception transition. But that feels out of scope, so I'm leaving it here as an observation alone.

**Review by shevek-google — COMMENTED** [review]
I think this is good. I had observations for the curious.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java (hunk 2 above)
Comment (aleofreddi) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java:136]: Good idea - done.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java (hunk 1 above)
Comment (aleofreddi) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java:147]: Done.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java (hunk 3 above)
Comment (aleofreddi) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java:204]: The for loop body can throw `IOException` and `JsonProcessingException` - migrating to a `forEach` method (in fact, we could also reuse Iterator's `forEachRemaining`) would need another functional interface that mimics `Consumer` allowing it to throw, or adding try/catch exceptions in place.

Given that such a functional interface would be used only once and that the current form is quite simple to read, I prefer to keep it as is.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerPageIterator.java [hunk 4]
```diff
@@ -40,13 +41,19 @@ static Page create(int offset, int limit) {
     abstract int limit();
   }
 
-  private final Function<Page, List<T>> fetcher;
+  @FunctionalInterface
+  public interface RangerFetcher<T> {
+
+    List<T> apply(Page page) throws RangerException;
```
Comment (paolomorandini) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerPageIterator.java:47]: Nit: consider naming this method `get` or `fetch`

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerPageIterator.java [hunk 5]
```diff
@@ -56,7 +63,12 @@ static Page create(int offset, int limit) {
   @Override
   protected T computeNext() {
     if (pageIterator == null || !pageIterator.hasNext() && !lastPage) {
-      List<T> entries = fetcher.apply(Page.create(offset, pageSize));
+      List<T> entries;
+      try {
+        entries = fetcher.apply(Page.create(offset, pageSize));
+      } catch (RangerException e) {
+        throw new RuntimeException("Failed to fetch data from Ranger", e);
```
Comment (paolomorandini) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerPageIterator.java:70]: Nit: consider including `offset` and `pageSize` in the exception message

**Review by paolomorandini — APPROVED** [review]
Thank you!

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerPageIterator.java (hunk 4 above)
Comment (aleofreddi) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerPageIterator.java:47]: Renamed to `fetch`, done.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerPageIterator.java (hunk 5 above)
Comment (aleofreddi) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerPageIterator.java:70]: Ok, done.
